File: templates/skills/workflow/scripts/path_utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
路径安全工具函数。

从 shared-utils.md 中提取的 resolveUnder 等安全关键逻辑。
安全逻辑不应依赖 AI 每次正确实现 — 必须由代码保证。
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Optional

WORKFLOW_STATE_FILENAME = "workflow-state.json"


# =============================================================================
# Public API
# =============================================================================


def resolve_under(base_dir: str, relative_path: str) -> Optional[str]:
    """安全解析相对路径，防止路径遍历攻击。

    返回绝对路径字符串，若路径不安全则返回 None。

    规则:
      - 拒绝空路径
      - 拒绝绝对路径
      - 拒绝包含 '..' 的路径
      - 拒绝包含特殊字符的路径
      - 拒绝解析后逃逸出 base_dir 的路径

    >>> resolve_under("/project", "src/main.py") is not None
    True
    >>> resolve_under("/project", "../etc/passwd") is None
    True
    >>> resolve_under("/project", "/etc/passwd") is None
    True
    """
    if not relative_path:
        return None

    if os.path.isabs(relative_path):
        return None

    if ".." in relative_path:
        return None

    # Only allow safe characters
    if not re.fullmatch(r"[a-zA-Z0-9_\-./]+", relative_path):
        return None

    # Reject leading slash, double slash, trailing whitespace-slash
    if re.search(r"^/|//|/\s*$", relative_path):
        return None

    resolved = os.path.realpath(os.path.join(base_dir, relative_path))
    normalized_base = os.path.realpath(base_dir)

    if resolved != normalized_base and not resolved.startswith(
        normalized_base + os.sep
    ):
        return None

    return resolved


def validate_project_id(project_id: str) -> bool:
    """验证项目 ID 是否安全（防注入）。

    >>> validate_project_id("abc123")
    True
    >>> validate_project_id("abc/../../etc")
    False
    >>> validate_project_id("")
    False
    """
    if not project_id:
        return False
    return bool(re.fullmatch(r"[a-zA-Z0-9_\-]+", project_id))


def get_workflows_dir(project_id: str) -> Optional[str]:
    """获取工作流目录路径。

    返回 ``~/.claude/workflows/{projectId}/`` 的绝对路径。
    若项目 ID 不安全则返回 None。
    """
    if not validate_project_id(project_id):
        return None

    home = Path.home()
    return str(home / ".claude" / "workflows" / project_id)


def get_workflow_state_path(project_id: str) -> Optional[str]:
    """获取 workflow-state.json 的 canonical 绝对路径。"""
    workflows_dir = get_workflows_dir(project_id)
    if not workflows_dir:
        return None
    return str(Path(workflows_dir) / WORKFLOW_STATE_FILENAME)


def is_canonical_workflow_state_path(state_path: str, project_id: Optional[str] = None) -> bool:
    """判断给定路径是否为合法的全局 workflow state 路径。"""
    if not state_path:
        return False

    candidate = Path(state_path).expanduser()
    if not candidate.is_absolute():
        return False

    resolved_path = candidate.resolve(strict=False)
    workflows_root_path = (Path.home() / ".claude" / "workflows").resolve(strict=False)
    resolved = str(resolved_path)
    workflows_root = str(workflows_root_path)
    if not resolved.startswith(workflows_root + os.sep):
        return False

    if resolved_path.name != WORKFLOW_STATE_FILENAME:
        return False

    parent = resolved_path.parent
    try:
        relative_parts = parent.relative_to(workflows_root_path).parts
    except ValueError:
        return False
    if len(relative_parts) != 1:
        return False

    detected_project_id = relative_parts[0]
    if not validate_project_id(detected_project_id):
        return False

    canonical = get_workflow_state_path(project_id or detected_project_id)
    if not canonical:
        return False
    return resolved == str(Path(canonical).resolve(strict=False))


def assert_canonical_workflow_state_path(state_path: str, project_id: Optional[str] = None) -> str:
    """校验 workflow state 路径必须是全局 canonical 路径。"""
    if not is_canonical_workflow_state_path(state_path, project_id):
        raise ValueError(
            "workflow-state.json must be stored under ~/.claude/workflows/{projectId}/workflow-state.json; "
            "project-local .claude/workflow-state.json is forbidden"
        )
    return str(Path(state_path).expanduser().resolve(strict=False))


def detect_project_id_from_root(project_root: Optional[str] = None) -> Optional[str]:
    """从项目根目录读取并校验 project-config.json 中的项目 ID。"""
    root = Path(project_root) if project_root else Path.cwd()
    config_path = root / ".claude" / "config" / "project-config.json"
    if not config_path.is_file():
        return None
    try:
        import json

        config = json.loads(config_path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None

    project = config.get("project") or {}
    project_id = project.get("id") or config.get("projectId")
    if not validate_project_id(str(project_id or "")):
        return None
    return str(project_id)


# =============================================================================
# CLI Entry
# =============================================================================


def main() -> int:
    import argparse
    import json
    import sys

    parser = argparse.ArgumentParser(description="路径安全工具")
    sub = parser.add_subparsers(dest="command")

    # resolve
    p_resolve = sub.add_parser("resolve", help="安全解析相对路径")
    p_resolve.add_argument("base_dir", help="基础目录")
    p_resolve.add_argument("relative_path", help="相对路径")

    # validate-id
    p_vid = sub.add_parser("validate-id", help="验证项目 ID")
    p_vid.add_argument("project_id", help="项目 ID")

    # workflows-dir
    p_wdir = sub.add_parser("workflows-dir", help="获取工作流目录路径")
    p_wdir.add_argument("project_id", help="项目 ID")

    # workflow-state-path
    p_state = sub.add_parser("workflow-state-path", help="获取状态文件路径")
    p_state.add_argument("project_id", help="项目 ID")

    # validate-state-path
    p_vstate = sub.add_parser("validate-state-path", help="校验状态文件路径")
    p_vstate.add_argument("state_path", help="状态文件绝对路径")
    p_vstate.add_argument("--project-id", help="可选项目 ID")

    args = parser.parse_args()

    if args.command == "resolve":
        result = resolve_under(args.base_dir, args.relative_path)
        print(json.dumps({"resolved": result}))
    elif args.command == "validate-id":
        result = validate_project_id(args.project_id)
        print(json.dumps({"valid": result}))
    elif args.command == "workflows-dir":
        result = get_workflows_dir(args.project_id)
        print(json.dumps({"path": result}))
    elif args.command == "workflow-state-path":
        result = get_workflow_state_path(args.project_id)
        print(json.dumps({"path": result}))
    elif args.command == "validate-state-path":
        result = is_canonical_workflow_state_path(args.state_path, args.project_id)
        print(json.dumps({"valid": result}))
    else:
        parser.print_help()
        return 1

    return 0


if __name__ == "__main__":
    import sys

    sys.exit(main())

File: core/utils/workflow/path_utils.js
#!/usr/bin/env node

const fs = require('fs')
const os = require('os')
const path = require('path')

const WORKFLOW_STATE_FILENAME = 'workflow-state.json'
const THINKING_GUIDES_DISPLAY_PATH = '.claude/.agent-workflow/specs/guides'
const THINKING_GUIDES_MANAGED_SEGMENTS = ['.claude', '.agent-workflow', 'specs', 'guides']
const THINKING_GUIDES_LEGACY_SEGMENTS = ['.claude', 'specs', 'guides']
const CODE_SPECS_DIR_SEGMENTS = ['.claude', 'code-specs']

function resolveUnder(baseDir, relativePath) {
  if (!relativePath) return null
  if (path.isAbsolute(relativePath)) return null
  if (relativePath.includes('..')) return null
  if (!/^[a-zA-Z0-9_./-]+$/.test(relativePath)) return null
  if (/^\/|\/\/|\/\s*$/.test(relativePath)) return null

  const resolved = path.resolve(baseDir, relativePath)
  const normalizedBase = path.resolve(baseDir)
  if (resolved !== normalizedBase && !resolved.startsWith(`${normalizedBase}${path.sep}`)) return null
  return resolved
}

function validateProjectId(projectId) {
  return /^[a-zA-Z0-9_-]+$/.test(String(projectId || ''))
}

function getWorkflowsDir(projectId) {
  if (!validateProjectId(projectId)) return null
  return path.join(os.homedir(), '.claude', 'workflows', projectId)
}

function getWorkflowStatePath(projectId) {
  const workflowsDir = getWorkflowsDir(projectId)
  return workflowsDir ? path.join(workflowsDir, WORKFLOW_STATE_FILENAME) : null
}

const HANDOFF_PHASES = ['spec', 'plan', 'execute']

function getHandoffDir(projectId) {
  const workflowsDir = getWorkflowsDir(projectId)
  return workflowsDir ? path.join(workflowsDir, 'handoff') : null
}

// handoff 落 handoff/{from-phase}.md（不入 state schema，覆盖式写）。
// fromPhase 仅允许 spec|plan|execute，非白名单或非法 projectId → null（不抛）。
function getHandoffPath(projectId, fromPhase) {
  if (!HANDOFF_PHASES.includes(fromPhase)) return null
  const handoffDir = getHandoffDir(projectId)
  return handoffDir ? path.join(handoffDir, `${fromPhase}.md`) : null
}

function isCanonicalWorkflowStatePath(statePath, projectId) {
  if (!statePath) return false
  const candidate = path.resolve(statePath)
  const workflowsRoot = path.resolve(path.join(os.homedir(), '.claude', 'workflows'))
  if (!candidate.startsWith(`${workflowsRoot}${path.sep}`)) return false
  if (path.basename(candidate) !== WORKFLOW_STATE_FILENAME) return false
  const relative = path.relative(workflowsRoot, path.dirname(candidate)).split(path.sep)
  if (relative.length !== 1) return false
  const detectedProjectId = relative[0]
  if (!validateProjectId(detectedProjectId)) return false
  const canonical = getWorkflowStatePath(projectId || detectedProjectId)
  return Boolean(canonical && candidate === path.resolve(canonical))
}

function assertCanonicalWorkflowStatePath(statePath, projectId) {
  if (!isCanonicalWorkflowStatePath(statePath, projectId)) {
    throw new Error('workflow-state.json must be stored under ~/.claude/workflows/{projectId}/workflow-state.json; project-local .claude/workflow-state.json is forbidden')
  }
  return path.resolve(statePath)
}

function readProjectConfig(projectRoot) {
  const root = projectRoot ? path.resolve(projectRoot) : path.resolve(normalizeWindowsShellPath(process.cwd()))
  const configPath = path.join(root, '.claude', 'config', 'project-config.json')
  try {
    return JSON.parse(fs.readFileSync(configPath, 'utf8'))
  } catch {
    return null
  }
}

function detectProjectIdFromRoot(projectRoot) {
  const config = readProjectConfig(projectRoot)
  if (!config) return null
  const project = config.project || {}
  const projectId = project.id || config.projectId
  return validateProjectId(projectId) ? String(projectId) : null
}

/**
 * 在 Windows 平台上把 Unix 风格的 shell 路径归一化为原生 Windows 路径。
 *
 *   /c/Users/...           → C:\Users\...   (Git Bash / MSYS2)
 *   /cygdrive/c/Users/...  → C:\Users\...   (Cygwin)
 *   /mnt/c/Users/...       → C:\Users\...   (WSL 路径泄漏)
 *
 * 已是 Windows 原生路径或不匹配的形态：直通（保守原则）。
 *
 * 必要性：Node 的 path.resolve 在 Windows 看到 `/d/xxx` 会前置当前驱动器（`D:\d\xxx`），
 * 导致 hook 找不到 `.claude/config/...`。
 */
const SHELL_DRIVE_RE = /^\/(?:cygdrive\/|mnt\/)?([A-Za-z])\/(.*)$/

function normalizeWindowsShellPath(p) {
  if (typeof p !== 'string' || !p) return p
  if (process.platform !== 'win32') return p
  const trimmed = p.trim()
  if (/^[A-Za-z]:[\/\\]/.test(trimmed)) return trimmed
  const m = SHELL_DRIVE_RE.exec(trimmed)
  if (!m) return trimmed
  return `${m[1].toUpperCase()}:\\${m[2].replace(/\//g, '\\')}`
}

function getThinkingGuidesDir(projectRoot, options = {}) {
  const { allowLegacy = true } = options
  const root = projectRoot ? path.resolve(projectRoot) : path.resolve(normalizeWindowsShellPath(process.cwd()))
  const managedDir = path.join(root, ...THINKING_GUIDES_MANAGED_SEGMENTS)
  if (fs.existsSync(managedDir) && fs.statSync(managedDir).isDirectory()) {
    return {
      path: managedDir,
      displayPath: THINKING_GUIDES_DISPLAY_PATH,
      source: 'managed',
    }
  }

  if (!allowLegacy) return null

  const legacyDir = path.join(root, ...THINKING_GUIDES_LEGACY_SEGMENTS)
  if (fs.existsSync(legacyDir) && fs.statSync(legacyDir).isDirectory()) {
    return {
      path: legacyDir,
      displayPath: THINKING_GUIDES_DISPLAY_PATH,
      source: 'legacy',
    }
  }

  return null
}

function getCodeSpecsDir(projectRoot) {
  const root = projectRoot ? path.resolve(projectRoot) : path.resolve(normalizeWindowsShellPath(process.cwd()))
  const dir = path.join(root, ...CODE_SPECS_DIR_SEGMENTS)
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) {
    return { path: dir, exists: false }
  }
  try {
    const resolvedRoot = fs.realpathSync(root)
    const resolvedDir = fs.realpathSync(dir)
    const expectedPrefix = path.join(resolvedRoot, ...CODE_SPECS_DIR_SEGMENTS)
    // 必须严格位于 <root>/.claude/code-specs 下，拒绝符号链接指向仓库根或仓库外部
    if (resolvedDir !== expectedPrefix && !resolvedDir.startsWith(expectedPrefix + path.sep)) {
      return { path: dir, exists: false }
    }
    return { path: resolvedDir, exists: true, resolvedRoot, expectedPrefix }
  } catch {
    return { path: dir, exists: false }
  }
}

function safeReadJson(targetPath, fallback = null, onError = null) {
  try {
    return JSON.parse(fs.readFileSync(targetPath, 'utf8'))
  } catch (err) {
    if (err && err.code === 'ENOENT') return fallback
    if (typeof onError === 'function') {
      try { return onError(err, targetPath) } catch { return fallback }
    }
    return fallback
  }
}

function main() {
  const [command, ...args] = process.argv.slice(2)
  if (command === 'resolve') {
    process.stdout.write(`${JSON.stringify({ resolved: resolveUnder(args[0], args[1]) })}\n`)
    return
  }
  if (command === 'validate-id') {
    process.stdout.write(`${JSON.stringify({ valid: validateProjectId(args[0]) })}\n`)
    return
  }
  if (command === 'workflows-dir') {
    process.stdout.write(`${JSON.stringify({ path: getWorkflowsDir(args[0]) })}\n`)
    return
  }
  if (command === 'workflow-state-path') {
    process.stdout.write(`${JSON.stringify({ path: getWorkflowStatePath(args[0]) })}\n`)
    return
  }
  if (command === 'validate-state-path') {
    const projectId = args[2] === '--project-id' ? args[3] : undefined
    process.stdout.write(`${JSON.stringify({ valid: isCanonicalWorkflowStatePath(args[0], projectId) })}\n`)
    return
  }
  process.stderr.write('Usage: node path_utils.js <resolve|validate-id|workflows-dir|workflow-state-path|validate-state-path> ...\n')
  process.exitCode = 1
}

module.exports = {
  WORKFLOW_STATE_FILENAME,
  resolveUnder,
  validateProjectId,
  getWorkflowsDir,
  getWorkflowStatePath,
  getHandoffDir,
  getHandoffPath,
  isCanonicalWorkflowStatePath,
  assertCanonicalWorkflowStatePath,
  detectProjectIdFromRoot,
  readProjectConfig,
  getThinkingGuidesDir,
  THINKING_GUIDES_DISPLAY_PATH,
  getCodeSpecsDir,
  CODE_SPECS_DIR_SEGMENTS,
  safeReadJson,
  normalizeWindowsShellPath,
}

if (require.main === module) main()
